kotlinx-coroutines-core/native/test/MultithreadedDispatchersTest.kt

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2016-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines
+
+import kotlin.native.concurrent.*
+import kotlin.test.*
+
+class MultithreadedDispatchersTest {
+    /**
+     * Test that [newFixedThreadPoolContext] does not allocate more dispatchers than it needs to.
+     * Incidentally also tests that it will allocate enough workers for its needs. Otherwise, the test will hang.
+     */
+    @Test
+    fun testNotAllocatingExtraDispatchers() {
+        suspend fun spin(set: MutableSet<Worker>) {
+            repeat(100) {
+                set.add(Worker.current)
+                delay(1)
+            }
+        }
+        val dispatcher = newFixedThreadPoolContext(64, "test")
+        try {
+            runBlocking {
+                val encounteredWorkers = mutableSetOf<Worker>()
+                var canStart1 = false
+                var canStart2 = false
+                val coroutine1 = launch(dispatcher) {
+                    while (!canStart1) {
+                        // intentionally empty
+                    }
+                    canStart2 = true
+                    spin(encounteredWorkers)
+                }
+                val coroutine2 = launch(dispatcher) {
+                    canStart1 = true
+                    while (!canStart2) {
+                        // intentionally empty
+                    }
+                    spin(encounteredWorkers)
+                }
+                listOf(coroutine1, coroutine2).joinAll()
+                assertEquals(2, encounteredWorkers.size)
+            }
+        } finally {
+            dispatcher.close()
+        }
```

Comment:
This test failed in the CI with `encounteredWorkers.size == 1`, but I fail to see how it's possible. Can different workers be equal to each other?

Reply:
It seems like there are at least two data-races:

* One is over `canStart1`/`canStart2`
* Another one is concurrent mutation of non-thread safe `MutableSet` which probably causes the failure

Reply:
> non-thread-safe `MultiSet`

Wow, what a ridiculous mistake. Looks like I've exhausted my resource for the understanding of concurrency for this PR on the main code.

> `canStart1`/`canStart2`

I still fail to see the problem, sorry, could you spell it out?

Reply:
We all have been there :)

>I still fail to see the problem, sorry, could you spell it out?

Non-atomic variable is read and write concurrently, I honestly not sure what guarantees K/N (or LLVM) provides in such scenarios. In theory, the compiler is allowed to host memory read in `while (!canStart*)` from the loop